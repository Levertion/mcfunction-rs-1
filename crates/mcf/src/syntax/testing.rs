use super::{group::McGroupType, lexer, tokens::McTokenKind, McParser, McfLang};
use mcfunction_parse::{parser::Parser, Ast, AstView, SyntaxKind, Token};

use std::fmt::Write;

pub fn format_astnode(node: AstView<&str, McfLang>, indlevel: usize) -> String {
    let ind = "    ".repeat(indlevel);
    let mut out = String::new();
    match node.kind() {
        SyntaxKind::Token(tk) => {
            if let McTokenKind::Whitespace = tk {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string().escape_debug(),
                    node.span()
                )
                .unwrap();
            } else {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string(),
                    node.span()
                )
                .unwrap();
            }
        }
        SyntaxKind::Group(gt) => {
            write!(
                out,
                "{}Group({}) at {} {{\n",
                ind,
                match gt {
                    McGroupType::CommandNode(ind) => format!("CommandNode({})", usize::from(*ind)),
                    v => format!("{:?}", v),
                },
                node.span()
            )
            .unwrap();
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel + 1)).unwrap();
            }
            write!(out, "{}}}\n", ind).unwrap();
        }
        SyntaxKind::Joined(gt) => {
            write!(
                out,
                "{}Joined({:?}) `{}` at {}\n",
                ind,
                gt,
                node.string(),
                node.span()
            )
            .unwrap();
            for x in node.children() {
                if let SyntaxKind::Error(err) = x.kind() {
                    write!(out, "{}- Error `{}`\n", ind, err).unwrap()
                }
            }
        }
        SyntaxKind::Error(err) => {
            write!(out, "{}Error `{}` at {}\n", ind, err, node.span()).unwrap();
        }
        SyntaxKind::Root(kind) => {
            for x in node.children() {
                write!(out, "Root({:?})\n{}", kind, format_astnode(x, indlevel)).unwrap();
            }
        }
    };
    out
}

pub fn format_sk_list(tokens: Vec<Vec<Token<McTokenKind>>>, src: &str) -> String {
    let mut out = String::new();
    for line in tokens {
        for tk in line {
            match tk.kind() {
                McTokenKind::Whitespace => write!(
                    out,
                    "{:?} `{}` at {}\n",
                    tk.kind(),
                    tk.string(src).escape_debug(),
                    tk.span(),
                ),
                McTokenKind::Eof if tk.start() != tk.end() => write!(
                    out,
                    "LineBreak `{}` at {}\n",
                    tk.string(src).escape_debug(),
                    tk.span(),
                ),
                _ => write!(
                    out,
                    "{:?} `{}` at {}\n",
                    tk.kind(),
                    tk.string(src),
                    tk.span(),
                ),
            }
            .unwrap();
        }
    }
    out
}

pub fn parse<'a, F: FnMut(&mut McParser)>(i: &'a str, mut f: F) -> Ast<&'a str, McfLang> {
    let tokens = lexer::tokenize_str(i);
    assert!(!tokens.is_empty(), "Token stream is empty");
    let mut parser = Parser::new(&tokens[0], i, McGroupType::File, false);
    f(&mut parser);
    for line in &tokens[1..] {
        parser.change_tokens(&line);
        f(&mut parser);
    }
    parser.build(false)
}
