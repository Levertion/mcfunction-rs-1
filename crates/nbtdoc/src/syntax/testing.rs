// I might need some of these in the future when I decide
// to write parser unit tests
#![allow(dead_code)]

use super::{group::NdGroupType, lexer, tokens::NdTokenKind, NbtdocLang, NdParser};
use mcfunction_parse::{ast::AstView, parser::Parser, Ast, SyntaxKind, Token};

use std::fmt::Write;

pub fn format_astnode(node: AstView<&str, NbtdocLang>, indlevel: usize) -> String {
    let ind = "    ".repeat(indlevel);
    let mut out = String::new();
    match node.kind() {
        SyntaxKind::Token(tk) => match tk {
            NdTokenKind::Whitespace => write!(
                out,
                "{}Token({:?}) `{}` at {}\n",
                ind,
                tk,
                node.string().escape_debug(),
                node.span()
            )
            .unwrap(),
            NdTokenKind::DocComment | NdTokenKind::Comment => write!(
                out,
                "{}Token({:?}) `{}` at {}\n",
                ind,
                tk,
                node.string().trim(),
                node.span()
            )
            .unwrap(),
            _ => write!(
                out,
                "{}Token({:?}) `{}` at {}\n",
                ind,
                tk,
                node.string(),
                node.span()
            )
            .unwrap(),
        },
        SyntaxKind::Group(gt) => {
            write!(out, "{}Group({:?}) at {} {{\n", ind, gt, node.span()).unwrap();
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel + 1)).unwrap();
            }
            write!(out, "{}}}\n", ind).unwrap();
        }
        SyntaxKind::Joined(gt) => {
            write!(
                out,
                "{}Joined({:?}) `{}` at {}\n",
                ind,
                gt,
                node.string(),
                node.span()
            )
            .unwrap();
            for x in node.children() {
                if let SyntaxKind::Error(err) = x.kind() {
                    write!(out, "{}- Error `{}`\n", ind, err).unwrap()
                }
            }
        }
        SyntaxKind::Error(err) => {
            write!(out, "{}Error `{}` at {}\n", ind, err, node.span()).unwrap();
        }
        SyntaxKind::Root(kind) => {
            for x in node.children() {
                write!(out, "Root({:?})\n{}", kind, format_astnode(x, indlevel)).unwrap();
            }
        }
    };
    out
}

pub fn format_sk_list(tokens: Vec<Token<NdTokenKind>>, src: &str) -> String {
    let mut out = String::new();
    for tk in tokens {
        match tk.kind() {
            NdTokenKind::Whitespace => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src).escape_debug(),
                tk.span(),
            ),
            NdTokenKind::DocComment | NdTokenKind::Comment => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src).trim(),
                tk.span()
            ),
            _ => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src),
                tk.span(),
            ),
        }
        .unwrap();
    }
    out
}

pub fn parse<'a, F: FnMut(&mut NdParser)>(i: &'a str, mut f: F) -> Ast<&'a str, NbtdocLang> {
    let tokens = lexer::tokenize_str(i);
    assert!(!tokens.is_empty(), "Token stream is empty");
    let mut parser = Parser::new(&tokens, i, NdGroupType::File, true);
    f(&mut parser);
    parser.build(false)
}
