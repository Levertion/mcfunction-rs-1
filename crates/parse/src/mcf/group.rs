use util::commands::Index;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum McGroupType {
    File,
    CommandNode(Index),
    Command,
    // NBT groups
    NbtCompound,
    NbtSequence,
    NbtNumber,
    NbtString,
    NbtBoolean,
    // NBT Keywords
    NbtSuffixB,
    NbtSuffixS,
    NbtSuffixL,
    NbtSuffixF,
    NbtSuffixD,
    NbtPrefixB,
    NbtPrefixI,
    NbtPrefixL,
    // NBT Compound
    NbtCompoundEntry,

    NbtPath,
    NbtPathSegment,
    NbtPathIndex,

    // Selector groups
    Selector,
    SelectorArgument,
    SelectorArgumentEntry,
    SelectorArgumentMap,
    SelectorArgumentMapEntry,
    // Selector keywords
    SelectorModP,
    SelectorModA,
    SelectorModR,
    SelectorModS,
    SelectorModE,

    // Block states
    BlockState,
    BlockStateArguments,

    ItemStack,
    ItemPredicate,

    Comment,

    Function,

    JsonObject,
    JsonObjectEntry,
    JsonList,
    JsonNull,

    Integer,
    Float,
    UnquotedString,
    ResourceLocation,
    Range,
    Uuid,
    Time,

    TimeS,
    TimeT,
    TimeD,

    Coord,
    CoordPart,

    // Keywords
    BooleanTrue,
    BooleanFalse,

    FloatSciExpUpper,
    FloatSciExpLower,

    // Special error type that signifies an error
    Error,
}
