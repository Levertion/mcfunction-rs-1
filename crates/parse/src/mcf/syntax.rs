use crate::{syntax::TokenKind, TokenSet};
use std::fmt;

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McTokenKind {
    // Single Character Punctuation
    Comma = 0,
    Dot,
    Colon,
    Semicolon,
    At,
    Excl,
    Eq,
    Lt,
    Gt,
    Slash,
    Tilde,
    Caret,
    Plus,
    Dash,
    Hash,
    // Double Character Punctuation
    DotDot,
    Lte,
    Gte,
    Swap,
    AddAssign,
    SubAssign,
    MulAssign,
    DivAssign,
    ModAssign,
    // Delimiters
    LBracket,
    RBracket,
    LCurly,
    RCurly,
    // Arbitrary Length
    QuotedString, // Does not have to be terminated, and will end at the end of any line
    Word,         // String of letters or underscores
    Whitespace,   // Non linebreak whitespace
    LineBreak,    // LF, CRLF, CR
    Digits,
    // Other
    Invalid, // Used to mark a lexing error
    Eof,     // Used to mark the end of a file
}

impl TokenKind for McTokenKind {
    const WHITESPACE: Self = Self::Whitespace;
    const EOF: Self = Self::Eof;
    const LINE_BREAK: Self = Self::LineBreak;
    const WORD: Self = Self::Word;
    const DELIMITERS: TokenSet<Self> = TokenSet::empty();
}

impl From<McTokenKind> for u8 {
    fn from(it: McTokenKind) -> Self {
        it as u8
    }
}

impl fmt::Display for McTokenKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use McTokenKind::*;

        write!(
            f,
            "{}",
            match self {
                Comma => ",",
                Dot => ".",
                Colon => ":",
                Semicolon => ";",
                At => "@",
                Excl => "!",
                Eq => "=",
                Lt => "<",
                Gt => ">",
                Hash => "#",
                Slash => "/",
                Tilde => "~",
                Caret => "^",
                Plus => "+",
                Dash => "-",
                DotDot => "..",
                Lte => "<=",
                Gte => ">=",
                Swap => "><",
                AddAssign => "+=",
                SubAssign => "-=",
                MulAssign => "*=",
                DivAssign => "/=",
                ModAssign => "%=",
                LBracket => "[",
                RBracket => "]",
                LCurly => "{",
                RCurly => "}",
                QuotedString => "Quoted String",
                Word => "Word",
                Whitespace => "Whitespace",
                LineBreak => "Linebreak",
                Digits => "Digits",
                Invalid => "Invalid",
                Eof => "EOF",
            }
        )
    }
}
