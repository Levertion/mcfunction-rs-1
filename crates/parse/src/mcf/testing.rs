use crate::{
    ast::{AstNode, SyntaxKind},
    mcf::{McGroupType, McTokenKind, McfLang},
    Token,
};

use std::fmt::Write;

pub fn format_astnode(node: &AstNode<McfLang>, indlevel: usize) -> String {
    let ind = "    ".repeat(indlevel);
    let mut out = String::new();
    match node.kind() {
        SyntaxKind::Token(tk) => {
            if let McTokenKind::Whitespace | McTokenKind::LineBreak = tk {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string().escape_debug(),
                    node.span()
                )
                .unwrap();
            } else {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string(),
                    node.span()
                )
                .unwrap();
            }
        }
        SyntaxKind::Group(gt) => {
            write!(
                out,
                "{}Group({}) at {} {{\n",
                ind,
                match gt {
                    McGroupType::CommandNode(ind) => format!("CommandNode({})", usize::from(*ind)),
                    v => format!("{:?}", v),
                },
                node.span()
            )
            .unwrap();
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel + 1)).unwrap();
            }
            write!(out, "{}}}\n", ind).unwrap();
        }
        SyntaxKind::Joined(gt) => {
            write!(
                out,
                "{}Joined({:?}) `{}` at {}\n",
                ind,
                gt,
                node.string(),
                node.span()
            )
            .unwrap();
            for x in node.children() {
                if let SyntaxKind::Error(err) = x.kind() {
                    write!(out, "{}- Error `{}`\n", ind, err).unwrap()
                }
            }
        }
        SyntaxKind::Error(err) => {
            write!(out, "{}Error `{}` at {}\n", ind, err, node.span()).unwrap();
        }
        SyntaxKind::Root => {
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel)).unwrap();
            }
        }
    };
    out
}

pub fn format_sk_list(tokens: Vec<Token<McTokenKind>>, src: &str) -> String {
    let mut out = String::new();
    for tk in tokens {
        match tk.kind() {
            McTokenKind::Whitespace | McTokenKind::LineBreak => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src).escape_debug(),
                tk.span(),
            ),
            _ => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src),
                tk.span(),
            ),
        }
        .unwrap();
    }
    out
}
