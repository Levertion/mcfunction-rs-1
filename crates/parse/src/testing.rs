use crate::{
    ast::{AstNode, GroupType, SyntaxKind},
    Token, TokenKind,
};

use std::fmt::Write;

pub fn format_astnode(node: &AstNode, indlevel: usize) -> String {
    let ind = "    ".repeat(indlevel);
    let mut out = String::new();
    match node.kind() {
        SyntaxKind::Token(tk) => {
            if let TokenKind::Whitespace | TokenKind::LineBreak = tk {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string().escape_debug(),
                    node.span()
                )
                .unwrap();
            } else {
                write!(
                    out,
                    "{}Token({:?}) `{}` at {}\n",
                    ind,
                    tk,
                    node.string(),
                    node.span()
                )
                .unwrap();
            }
        }
        SyntaxKind::Group(gt) => {
            write!(
                out,
                "{}Group({}) at {} {{\n",
                ind,
                match gt {
                    GroupType::CommandNode(ind) => format!("CommandNode({})", usize::from(*ind)),
                    v => format!("{:?}", v),
                },
                node.span()
            )
            .unwrap();
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel + 1)).unwrap();
            }
            write!(out, "{}}}\n", ind).unwrap();
        }
        SyntaxKind::Joined(gt) => {
            write!(
                out,
                "{}Joined({:?}) `{}` at {}\n",
                ind,
                gt,
                node.string(),
                node.span()
            )
            .unwrap();
            for x in node.children() {
                if let SyntaxKind::Error(err) = x.kind() {
                    write!(out, "{}- Error `{}`\n", ind, err).unwrap()
                }
            }
        }
        SyntaxKind::Error(err) => {
            write!(out, "{}Error `{}` at {}\n", ind, err, node.span()).unwrap();
        }
        SyntaxKind::Root => {
            for x in node.children() {
                write!(out, "{}", format_astnode(x, indlevel)).unwrap();
            }
        }
    };
    out
}

pub fn format_sk_list(tokens: Vec<Token>, src: &str) -> String {
    let mut out = String::new();
    for tk in tokens {
        match tk.kind() {
            TokenKind::Whitespace | TokenKind::LineBreak => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src).escape_debug(),
                tk.span(),
            ),
            _ => write!(
                out,
                "{:?} `{}` at {}\n",
                tk.kind(),
                tk.string(src),
                tk.span(),
            ),
        }
        .unwrap();
    }
    out
}
