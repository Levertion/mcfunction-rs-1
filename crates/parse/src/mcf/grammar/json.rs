use super::*;
use crate::{
    mcf::{
        group::McGroupType::{self, *},
        syntax::McTokenKind::*,
        McParser,
    },
    parser::StartInfo::*,
    TokenSet,
};

const JSON_NULL: &[(&str, McGroupType)] = &[("null", JsonNull)];

pub fn object(p: &mut McParser) {
    let objmk = p.start(JsonObject, Skip);
    p.expect(LCurly);
    if p.eat(RCurly) {
        p.finish(objmk);
    } else {
        loop {
            let entmk = p.start(JsonObjectEntry, Skip);
            if p.expect(QuotedString) && p.expect(Colon) {
                value(p);
            }
            p.finish(entmk);
            if p.at(RCurly) {
                break;
            }
            match (p.expect(Comma), p.at(Eof)) {
                (true, true) => {
                    p.expect(QuotedString);
                    break;
                }
                (false, false) => {
                    if p.bump_recover(TokenSet::empty()) {
                        break;
                    }
                }
                (true, false) => (),
                (false, true) => {
                    break;
                }
            }
        }
        p.expect(RCurly);
        p.finish(objmk);
    }
}

pub fn array(p: &mut McParser) {
    let arrmk = p.start(JsonList, Skip);
    p.expect(LBracket);
    if p.eat(RBracket) {
        p.finish(arrmk);
    } else {
        loop {
            value(p);
            if p.at(RBracket) {
                break;
            }
            match (p.expect(Comma), p.at(Eof)) {
                (true, true) => {
                    value(p);
                    break;
                }
                (false, false) => {
                    if p.bump_recover(TokenSet::empty()) {
                        break;
                    }
                }
                (true, false) => (),
                (false, true) => {
                    break;
                }
            }
        }
        p.expect(RBracket);
        p.finish(arrmk);
    }
}

pub fn value(p: &mut McParser) {
    let mut lk = p.lookahead();
    if lk.at(LCurly) {
        object(p);
    } else if lk.at(LBracket) {
        array(p);
    } else if lk.at(QuotedString) {
        p.bump();
    } else if lk.at_keywords(BOOLEAN) {
        p.eat_keyword(BOOLEAN);
    } else if lk.at_keywords(JSON_NULL) {
        p.eat_keyword(JSON_NULL);
    } else {
        lk.group_error(Float);
        let errs = lk.get_errors();
        if !p.try_token(float_tk, Float) {
            p.add_errors(errs);
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::mcf::testing::{format_astnode, parse};

    use insta::assert_snapshot;

    macro_rules! json_test {
        ($name:ident, $e:expr) => {
            #[test]
            fn $name() {
                assert_snapshot!(parse_json($e));
            }
        };
    }

    fn parse_json(i: &str) -> String {
        format_astnode(&parse(i, super::value), 0)
    }

    json_test!(boolean_true, "true");
    json_test!(boolean_false, "false");
    json_test!(null, "null");
    json_test!(
        string,
        r#""hello world! This is a string\n with\n escapes""#
    );
    json_test!(object_empty, "{}");
    json_test!(object_simple, r#"{"text":"hello"}"#);
    json_test!(object_multi, r#"{"text":"hello","bold":true}"#);
    json_test!(object_nokey, r#"{"foo":true,}"#);
    json_test!(object_nocolon, r#"{"foo":true,"bar"}"#);
    json_test!(object_nokey_unclosed, r#"{"foo":true,"#);

    json_test!(array_empty, "[]");
    json_test!(array_simple, r#"[1, true, "hello"]"#);
    json_test!(array_noval, r#"[1, ]"#);
    json_test!(array_unclosed, r#"["hello", "#);
    json_test!(array_nocomma, r#"[true"#);
}
