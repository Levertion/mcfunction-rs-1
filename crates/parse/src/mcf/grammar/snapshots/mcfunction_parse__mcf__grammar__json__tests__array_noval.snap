---
source: crates/parse/src/mcf/grammar/json.rs
expression: "parse_json(r#\"[1, ]\"#)"
---
Group(JsonList) at 0:0 - 0:5 {
    Token(LBracket) `[` at 0:0 - 0:1
    Joined(Float) `1` at 0:1 - 0:2
    Token(Comma) `,` at 0:2 - 0:3
    Error `Expected one of '{', '[', 'Quoted String'` at 0:4 - 0:5
    Error `Expected one of 'true', 'false', 'null'` at 0:4 - 0:5
    Error `Error while parsing 'Float'` at 0:4 - 0:5
    Token(RBracket) `]` at 0:4 - 0:5
}

